logstreams/src/main/java/io/zeebe/distributedlog/restore/impl/RestoreController.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.distributedlog.restore.impl;
+
+import io.atomix.utils.concurrent.SingleThreadContext;
+import io.atomix.utils.concurrent.ThreadContext;
+import io.zeebe.distributedlog.restore.RestoreClient;
+import io.zeebe.distributedlog.restore.RestoreFactory;
+import io.zeebe.distributedlog.restore.RestoreNodeProvider;
+import io.zeebe.distributedlog.restore.RestoreStrategy;
+import io.zeebe.distributedlog.restore.log.LogReplicationAppender;
+import io.zeebe.distributedlog.restore.log.LogReplicator;
+import io.zeebe.logstreams.state.SnapshotRequester;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+
+public class RestoreController {
+
+  private final ThreadContext restoreThreadContext;
+  private final Logger logger;
+  private DefaultStrategyPicker strategyPicker;
+
+  public RestoreController(
+      int partitionId, LogReplicationAppender appender, RestoreFactory clientFactory, Logger log) {
+    this.restoreThreadContext =
```

Comment:
I'm also guilty of this :sweat_smile: but I think we should try to avoid constructing objects in a constructor, and aim to use IoC as much as possible, so in this case we should also receive the thread context and strategy picker as arguments. That last one though, we should discuss if it doesn't make sense to merge it into this controller, at which point it's just about receiving the thread context.

I'm also OK with having a constructor which takes in all member fields as arguments, and another one which reuses it to provide sane defaults, a bit like the `BrokerRestoreServer`. That's not 100% IoC of course, but I think it's a good compromise to keep the advantages of both world.